### PR TITLE
Improve handling and detection of region codes and names

### LINKED ISSUE #19359: France regions 20% of unknown region
When using the Maxmind database, 20% of the cities are not matched with a region, for example: Dinan, France - Deauville, France

### LINKED ISSUE #20527: Regions might not be detected correctly when using free DBIP geolocation database
## Context

When geolocation is enabled in Matomo, Matomo tries to detect and store the location of the visitor based on his IP.

Depending on the use location database, this may contain country, region and city details.

Countries and Regions are internally stored as iso codes.

Unlike most other databases the free DBIP location database does not include any region codes. Instead it only serves region names.
To be able to store the region details in that case, we try to look up the region code by it's name using a [list of all regions](https://github.com/matomo-org/matomo/blob/5.x-dev/plugins/GeoIp2/data/isoRegionNames.php).

This currently fails as in some cases our list is not up to date in other cases the database returns the name in e.g. another language or writing.

As the free DBIP geolocation database is the suggested default database, we should improve that detection/mapping.

This problem has already been reported in various other issues like: https://github.com/matomo-org/matomo/issues/19359, https://github.com/matomo-org/matomo/issues/19329, https://github.com/matomo-org/matomo/issues/19411, https://github.com/matomo-org/matomo/issues/16714

## Expected Behavior

Regions should be detected correctly when using free DBIP location database

## Current Behavior

Regions are only partially or not detected at all for certain countries.

## Possible Solution

The detection of regions should be done a bit more fuzzy. In addition the list should also contain old names of regions, regions that have been removed from iso and maybe also some other names that are used by DBIP database.

A proposed solution can be found in https://github.com/matomo-org/matomo/pull/20420

@mattab @Stan-vw Might be good to prioritize this one into Matomo 5 as it might impact a lot on premise customers.
The proposed solution is almost finished. Might need a couple more hours to look it through again and maybe add some more tests.

### PR DESCRIPTION
**Pull request by sgiehl**
### Summary:

This PR provides the following improvements & benefits around regions:

* Updated list of all iso regions known to Matomo
* Accurancy of region detection with DB IP Lite database is increased 
* Updating the list of iso regions will no longer cause revised regions to no longer be displayed correctly in older datasets

### Description:

The DB-IP lite database, which Matomo is currently using as default, does only provide region names in it's detection (no iso codes).
As Matomo only stores region codes in the database, we are trying to match the returned region name to the region names that Matomo knows about. This might fail if the region name returned by the geoip database doesn't match a known name, which actually seems to be the case quite often.

To improve the region detection, this PR adjusts the lists of known regions. Instead of only holding the latest set of iso region codes mapped to there current name, the array now looks like this:
```php
 <CountryCode> => [
     <RegionCode> => [
         'name' => <CurrentISOName>
         'altNames' => [
             // list of previous names or names used by GeoIP providers like db-ip
         ],
         'current' => <bool> indicating if the iso code is currently used
     ]
]
``` 

By providing a list of alternate names it will in the future increase the accurancy when trying to match regions by its name.

To properly fill this new strutuce the command to update the region list has been adjusted. It now also includes an additional option to add names used in db ip database, that don't match known region names.

The region list included in this PR was built with this steps:
* Initial creation of the region list using an older version of the iso-codes project (so older names and regions are included)
* Update the list with the latest revision of iso-codes (so we have the latest set as current, but still older names as alternate names)
* Enrich regions and alternate names with the latest version of db ip databases

fixes #20527
refs #20368

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: plugins/GeoIp2/Commands/UpdateRegionCodes.php [hunk 1]
```diff
@@ -128,4 +155,72 @@ protected function doExecute(): int
...
+        $supportedInfo = $php->getSupportedLocationInfo();
+
+        if (empty($supportedInfo[LocationProvider::REGION_CODE_KEY])) {
+            $output->writeln(' <fg=red>X Region codes not supported by currently used GeoIP database. Skipping.</>');
+            return;
+        }
+
+        $output->writeln('Iterating through all IPv4 addresses...');
+
+        $this->initProgressBar(6396645);
```
Comment (michalkleiner) [inline plugins/GeoIp2/Commands/UpdateRegionCodes.php:176]: Probably doesn't matter much but what is this number based on? Running the full CSV from MMDB import and the CSV has 42479761 lines, roughly one order more, so the progress bar shows 100% for over 95% of the time.

Comment (michalkleiner) [reply plugins/GeoIp2/Commands/UpdateRegionCodes.php:176]: Could be I have a different set.

**Review by michalkleiner — APPROVED** [review]
I wasn't able to fully revert the change to `isoRegionNames.php` and let it regenerate since the content of it and the code relating to it changed gradually.
Since the version of the file is part of the PR here I don't see that as an issue.

Back to you @sgiehl if you want to ask Ben to have a look or merge as is.

File: plugins/GeoIp2/Commands/UpdateRegionCodes.php (hunk 1 above)
Comment (sgiehl) [reply plugins/GeoIp2/Commands/UpdateRegionCodes.php:176]: That was the number of records in the lite database. It's enough to use the CSV file of the lite version, while using the MMDB file of the paid version in the backend. There is no need to iterate over all records of the paid version, as only the lite version lacks the region code...

**Pull request merged**